quaternions/general_quaternion.py

```diff
@@ -0,0 +1,185 @@
+    represents a quaternion, considered a member in Lie algebra of quaternions.
+    for backward compatibility shares the same interface with Quaternion.
+    unlike quaternion, it's not defined up to scale
+    """
+    def __init__(self, qr, qi, qj, qk):
+        self.qr = qr
+        self.qi = qi
+        self.qj = qj
+        self.qk = qk
+
+    def __add__(self, p):
+        if not is_quaternion(p):
+            raise QuaternionError('expected quaternion, got %s' % p)
+        return self.__class__(self.qr + p.qr, self.qi + p.qi, self.qj + p.qj, self.qk + p.qk)
+
+    def __sub__(self, p):
+        if not is_quaternion(p):
+            raise QuaternionError('expected quaternion, got %s' % p)
+        return self.__class__(self.qr - p.qr, self.qi - p.qi, self.qj - p.qj, self.qk - p.qk)
+
+    def __neg__(self):
+        return self.__class__(-self.qr, -self.qi, -self.qj, -self.qk)
+
+    def __mul__(self, p):
+        if is_quaternion(p):
+            mat = np.array([
+                [self.qr, -self.qi, -self.qj, -self.qk],  # noqa
+                [self.qi,  self.qr,  self.qk, -self.qj],  # noqa
+                [self.qj, -self.qk,  self.qr,  self.qi],  # noqa
+                [self.qk,  self.qj, -self.qi,  self.qr]   # noqa
+            ])
+            result = mat.dot(np.array(p.coordinates))
+            return self.__class__(*result)
+        else:
+            return self.__class__(self.qr * p, self.qi * p, self.qj * p, self.qk * p)
+
+    def __rmul__(self, p):
+        return self.__mul__(p)
+
+    def __truediv__(self, p):
+        return self * (1 / p)
+
+    def __rtruediv__(self, p):
+        return p * self.inverse()
+
+    def conjugate(self):
+        return self.__class__(self.qr, -self.qi, -self.qj, -self.qk)
+
+    def inverse(self):
+        return self.conjugate() * (1 / self._squarenorm())
+
+    def __invert__(self):
+        return self.inverse()
+
+    def _squarenorm(self):
+        return self.qr * self.qr + self.qi * self.qi + self.qj * self.qj + self.qk * self.qk
+
+    def __repr__(self):
+        return 'GeneralQuaternion{}'.format(self.coordinates)
+
+    def __str__(self):
+        return '({qr:.6g}{qi:+.6g}i{qj:+.6g}j{qk:+.6g}k)'.format(**self.__dict__)
+
+    def is_real(self, tolerance=DEFAULT_TOLERANCE):
+        """ True if i, j, k components are zero. """
+        complex_norm = np.linalg.norm([self.qi, self.qj, self.qk])
+        return complex_norm < tolerance
+
+    def is_equal(self, other, tolerance=DEFAULT_TOLERANCE):
+        """
+        compares as quaternions up to tolerance.
+        Note: tolerance in coords, not in quaternions metrics.
+        Note: unlike quaternions, equality is not up to scale.
+        """
+        return np.linalg.norm(self.coordinates - other.coordinates) < tolerance
+
+    def __eq__(self, other):
+        return self.is_equal(other)
+
+    def norm(self):
+        return np.sqrt(self._squarenorm())
+
+    def normalized(self):
+        return self / self.norm()
```

Comment:
i wasn't sure about it. i didn't want GeneralQuaternion to know about derived class, it's cyclic. what do you think?

Reply:
Agree, it is not the best thing to do.
Actually, the constructor for `Quaternion`  now normalizes quaternions, so one could do
`Quaternion(*q.coordinates)` and get the normalized quaternion with the unitary class.

Ok, let's leave it like this. We could add a comment about how to get a `Quaternion`  instead of a `GeneralQuaternion` here.